realm/realm-library/src/main/java/io/realm/RealmResults.java

```diff
@@ -56,22 +58,24 @@
  * @see Realm#executeTransaction(Realm.Transaction)
  */
 public class RealmResults<E extends RealmModel> extends OrderedRealmCollectionImpl<E> {
-    static <T extends RealmModel> RealmResults<T> createBacklinkResults(Realm realm, UncheckedRow row, Class<T> srcTableType, String srcFieldName) {
+    @SuppressLint("unused")
+    static <T extends RealmModel> RealmResults<T> createBacklinkResults(BaseRealm realm, Row row, Class<T> srcTableType, String srcFieldName) {
+        UncheckedRow uncheckedRow = (UncheckedRow) row;
         Table srcTable = realm.getSchema().getTable(srcTableType);
-        return new RealmResults<T>(
+        return new RealmResults<>(
                 realm,
-                Collection.createBacklinksCollection(realm.sharedRealm, row, srcTable, srcFieldName),
+                Collection.createBacklinksCollection(realm.sharedRealm, uncheckedRow, srcTable, srcFieldName),
                 srcTableType);
     }
 
-    static RealmResults<DynamicRealmObject> createBacklinkResults(DynamicRealm realm, CheckedRow row, Table srcTable, String srcFieldName) {
+    static RealmResults<DynamicRealmObject> createDynamicBacklinkResults(BaseRealm realm, Row row, Table srcTable, String srcFieldName) {
```

Comment:
Ok. I've reverted this.

I *hate* it.  In the call, we have to cast to get the parameter type right and then there is *nothing* inside the method that requires any of the more constrained types.  This is an abomination.  After looking at the code, though, I agree that it is better with the constrained params than it is without.

I think that the root problem is that there is no generic parameter on `DynamicObject`, so we lose all typing information as soon as we create one.

Out of scope.  Holding my nose.